File: solutions/cpp/permutations-alternative.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::stoul;
using std::swap;
using std::uniform_int_distribution;
using std::vector;

vector<vector<int>> PermutationsHelper(int i, vector<int> *A);

// @include
vector<vector<int>> Permutations(vector<int> A) {
  return PermutationsHelper(0, &A);
}

vector<vector<int>> PermutationsHelper(int i, vector<int> *A) {
  if (i == A->size() - 1) {
    return {*A};
  }

  vector<vector<int>> result;
  for (int j = i; j < A->size(); ++j) {
    swap((*A)[i], (*A)[j]);
    vector<vector<int>> new_result = PermutationsHelper(i + 1, A);
    result.insert(result.end(), new_result.begin(), new_result.end());
    swap((*A)[i], (*A)[j]);
  }
  return result;
}
// @exclude

template <typename T>
bool EqualVector(const vector<T> &A, const vector<T> &B) {
  return A.size() == B.size() && equal(A.begin(), A.end(), B.begin());
}

void SmallTest() {
  vector<int> A = {0, 1, 2};
  auto result = Permutations(A);
  assert(result.size() == 6);
  vector<vector<int>> golden_result = {{0, 1, 2}, {0, 2, 1}, {1, 0, 2}, {1, 2, 0}, {2, 1, 0}, {2, 0, 1}};
  for (int i = 0; i < 6; ++i) {
    assert(EqualVector(result[i], golden_result[i]));
  }
}

int main(int argc, char** argv) {
  SmallTest();
  default_random_engine gen((random_device())());
  int n;
  if (argc == 2) {
    n = stoul(argv[1]);
  } else {
    uniform_int_distribution<int> dis(1, 10);
    n = dis(gen);
  }
  vector<int> A(n);
  iota(A.begin(), A.end(), 0);
  auto result = Permutations(A);
  cout << "n = " << n << endl;
  for (const auto& vec : result) {
    for (int a : vec) {
      cout << a << " ";
    }
    cout << endl;
  }
  return 0;
}

File: static/solutions/java/src/main/java/com/epi/PermutationsAlternative.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class PermutationsAlternative {
  // @include
  public static List<List<Integer>> permutations(List<Integer> A) {
    List<List<Integer>> result = new ArrayList<>();
    directedPermutations(0, A, result);
    return result;
  }

  private static void directedPermutations(int i, List<Integer> A,
                                           List<List<Integer>> result) {
    if (i == A.size() - 1) {
      result.add(new ArrayList<>(A));
      return;
    }

    // Try every possibility for A[i].
    for (int j = i; j < A.size(); ++j) {
      Collections.swap(A, i, j);
      // Generate all permutations for A.subList(i + 1, A.size()).
      directedPermutations(i + 1, A, result);
      Collections.swap(A, i, j);
    }
  }
  // @exclude

  private static void smallTest() {
    List<Integer> A = Arrays.asList(0, 1, 2);
    List<List<Integer>> result = permutations(A);
    assert(result.size() == 6);
    List<List<Integer>> goldenResult = Arrays.asList(
        Arrays.asList(0, 1, 2), Arrays.asList(0, 2, 1), Arrays.asList(1, 0, 2),
        Arrays.asList(1, 2, 0), Arrays.asList(2, 1, 0), Arrays.asList(2, 0, 1));
    assert(result.equals(goldenResult));
  }

  public static void main(String[] args) {
    smallTest();
    Random r = new Random();
    int n;
    if (args.length == 1) {
      n = Integer.parseInt(args[0]);
    } else {
      n = r.nextInt(8) + 1;
    }
    List<Integer> A = new ArrayList<>(n);
    int val = 0;
    for (int i = 0; i < n; i++) {
      A.add(val++);
    }
    List<List<Integer>> result = permutations(A);
    System.out.println("n = " + n);
    for (List<Integer> vec : result) {
      System.out.println(vec);
    }
  }
}
